timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/HeldLocks.java

```diff
@@ -32,21 +36,32 @@
     private final Collection<AsyncLock> acquiredLocks;
     private final LockToken token;
     private final LeaseExpirationTimer expirationTimer;
+    private final LockWatchingService lockWatchingService;
+    private final Supplier<Set<LockDescriptor>> descriptors = Suppliers.memoize(this::getLockDescriptors);
 
     @GuardedBy("this")
     private boolean isUnlocked = false;
 
-    public HeldLocks(LockLog lockLog, Collection<AsyncLock> acquiredLocks, UUID requestId, LeaderClock leaderClock) {
-        this(lockLog, acquiredLocks, requestId, new LeaseExpirationTimer(() -> leaderClock.time().currentTime()));
-    }
-
     @VisibleForTesting
     HeldLocks(LockLog lockLog, Collection<AsyncLock> acquiredLocks,
-            UUID requestId, LeaseExpirationTimer expirationTimer) {
+            UUID requestId, LeaseExpirationTimer expirationTimer, LockWatchingService lockWatchingService) {
         this.lockLog = lockLog;
         this.acquiredLocks = acquiredLocks;
         this.token = LockToken.of(requestId);
         this.expirationTimer = expirationTimer;
+        this.lockWatchingService = lockWatchingService;
+    }
+
+    public static HeldLocks create(LockLog lockLog, Collection<AsyncLock> acquiredLocks, UUID requestId,
+            LeaderClock leaderClock, LockWatchingService lockWatchingService) {
+        HeldLocks locks = new HeldLocks(lockLog, acquiredLocks, requestId,
+                new LeaseExpirationTimer(() -> leaderClock.time().currentTime()), lockWatchingService);
+        locks.registerLock();
+        return locks;
```

Comment:
cool, so before a lock is visible, it's registered in the lock log. Also, before a lock is reaped it must first become visible, and we register unlock in the log only when unlocking. Thus, as far as each lock is concerned, locks/unlocks are linearizable between the log and the actual thing happening. The remainder of the correctness arguments are in the internal rfc.